chunkset.c

```diff
@@ -34,27 +34,24 @@ static inline void chunkmemset_1(uint8_t *from, chunk_t *chunk) {
 static inline void chunkmemset_4(uint8_t *from, chunk_t *chunk) {
 #if defined(UNALIGNED64_OK)
     uint32_t half_chunk;
-    half_chunk = *(uint32_t *)from;
+    memcpy(&half_chunk, from, sizeof(half_chunk));
     *chunk = 0x0000000100000001 * (uint64_t)half_chunk;
 #elif defined(UNALIGNED_OK)
-    chunk->u32[0] = *(uint32_t *)from;
+    memcpy(&chunk->u32[0], from, sizeof(chunk->u32[0]));
     chunk->u32[1] = chunk->u32[0];
 #else
     uint8_t *chunkptr = (uint8_t *)chunk;
-    memcpy(chunkptr, from, 4);
-    memcpy(chunkptr+4, from, 4);
+    memcpy(chunkptr, from, sizeof(uint32_t));
+    memcpy(chunkptr+4, from, sizeof(uint32_t));
 #endif
 }
 
 static inline void chunkmemset_8(uint8_t *from, chunk_t *chunk) {
-#if defined(UNALIGNED64_OK)
-    *chunk = *(uint64_t *)from;
-#elif defined(UNALIGNED_OK)
-    uint32_t* p = (uint32_t *)from;
-    chunk->u32[0] = p[0];
-    chunk->u32[1] = p[1];
+#if defined(UNALIGNED_OK) && !defined(UNALIGNED64_OK)
```

Comment:
I have been looking at this one too, is there any point to splitting this up into two memcpy calls or could we just use the single one below?
My limited testing does not show any change, but I have not dug into whether other `UNALIGNED_OK`-enabled code or endianness requires this somehow.

I do get the feeling that chunkset.c could use a refactoring after this though.

Reply:
I suppose that depends on if the compiler automatically converts 1 x 8 copy to 2 x 4 copies if the architecture doesn't support 64-bit unaligned copies.

Reply:
Hmm, I thought that was guaranteed to be handled by memcpy automatically, now I wonder

Reply:
I'm not exactly sure what hardware configuration this would apply to. I don't mind if `chunk_t` was only `uint64_t` and we got rid of the `ifndef UNALIGNED64_OK` paths. Perhaps that can be done as a separate PR..

Reply:
https://godbolt.org/z/sxq61bYPK

Reply:
![image](https://user-images.githubusercontent.com/1364432/148615174-78d07b67-d917-403a-9818-c95a73d262f1.png)

For ARM (32-bit) it requires at least GCC 9 in order to take advantage of it. But it does convert it to two 4 byte unaligned operations.

Reply:
memcpy is required to handle unaligned buffers btw.

Seems like GCC8 on ARM just calls the generic memcpy, but GCC9 uses its builtin handling.

From what I gather, just using memcpy should be perfectly safe (and probably more performant too).
I re-ran the tests with changing the below to always use the 8-byte copy, and that regains most of the performance loss too. Possibly GCC 4.8 issues two calls to the generic memcpy function.

I don't really think we need to special-case memcpy from what I am seeing so far.

Reply:
Here is another view:
![image](https://user-images.githubusercontent.com/1364432/148618510-2949cc33-cf2e-4405-9a90-cb9006bd2cd5.png)
You can see in the last compiler ARM GCC 4 that the first one calls `memcpy` and the second one doesn't.

Reply:
Yes, but I am not sure the performance loss is enough to warrant the workaround, in addition to it only happening on the older compilers AFAIK. If the alternative is unaligned type-punning I mean.

Reply:
Ok that makes sense. I can rework it in another PR after this one is merged. I will just remove the `typedef struct chunk_t { uint32_t u32[2]; } chunk_t;` and all its associated code as most modern compilers are able to do 8 byte `memcpy` optimizations themselves.

Reply:
I wonder if we should just also not do `zmemcmp` and only replace with `memcmp`. And we can just note these compiler oddies in the Wiki page somewhere. We do list recommended build minimums to be GCC 5.5. Then in CMake and Configure we could print a warning. Thoughts?

Reply:
@nmoinvaz I am keeping that possibility open for sure, but I want to do the benchmarks and such first so we have better knowledge about the downsides before deciding.